### PR TITLE
[ES-1060] Feature/optimize upsert empty update

### PR DESCRIPTION
**Pull request by cpjulia**
### Scope & Purpose

Added back the optimization for empty document update operations (i.e. update requests in which no attributes were specified to be updated), handling them in a special way without performing any writes, also excluding such special cases of operation from replication to followers.
Example: 
```
UPSERT {name: "test1"} INSERT {name: "test1"} UPDATE {} IN testCollection OPTIONS { } RETURN { new: NEW, old: OLD }
```
- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [x] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [x] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [x] Backports (Obs: If this doesn't produce unwanted side effects, it will be backported)
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*
  - [ ] Backport for 3.7: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [x] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/ES-1060
- [ ] Design document:

### CODE REVIEW

File: arangod/RocksDBEngine/RocksDBCollection.cpp [hunk 1]
```diff
@@ -1224,6 +1224,15 @@ Result RocksDBCollection::performUpdateOrReplace(
     }
   }
 
+  if (newSlice.length() <= 1 && isUpdate) {  // TODO move above ?!
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBCollection.cpp:1227]: The TODO is obsolete here.
Please also leave a comment why we are exiting early here, and why we are comparing the length of `newSlice` to 1 and not 0. That makes it easier in the future to understand this code part.
```suggestion
  if (newSlice.length() <= 1 && isUpdate) {
```

File: arangod/Transaction/Methods.cpp [hunk 2]
```diff
@@ -1510,28 +1514,39 @@ Future<OperationResult> transaction::Methods::modifyLocal(
...
     }
-
     // it is ok to clobber res here!
     res.reset();
   } else {
-    res = workForOneDocument(newValue, false);
+    bool excludeFromReplication = false;
+    res = workForOneDocument(newValue, false, excludeFromReplication);
+    if (res.ok() && excludeFromReplication) {
+      excludePositions.insert(0);
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp:1548]: Ok, `excludePositions` are populated here, but later on nothing happens with them.
Especially, **they are not passed to replicateOperations**, so they are effectively useless.

File: tests/js/server/aql/aql-modify-cluster.js [hunk 3]
```diff
@@ -69,11 +71,19 @@ function ahuacatlModifySuite () {
     setUp : function () {
       db._drop(cn1);
       db._drop(cn2);
+      db._drop(cn3);
       c1 = db._create(cn1, {numberOfShards:5});
       c2 = db._create(cn2, {numberOfShards:5});
+      c3 = db._create(cn3, {numberOfShards:5});
```
Comment (jsteemann) [inline tests/js/server/aql/aql-modify-cluster.js:77]: We are already creating 2 new collections for each test case.
Why do we need a 3rd one that is unused in most test cases, and used only in a few new ones?
Can't we just use `c1` or `c2` in the new test cases?

File: tests/js/server/aql/aql-modify-cluster.js [hunk 4]
```diff
@@ -1371,6 +1413,19 @@ function ahuacatlUpdateSuite () {
       }
     },
 
+///////////////////////////////////////////////////////////////////////////////
+/// @brief test upsert with empty object
```
Comment (jsteemann) [inline tests/js/server/aql/aql-modify-cluster.js:1417]:
```suggestion
/// @brief test update with empty object
```

File: tests/js/server/aql/aql-modify-noncluster.js [hunk 5]
```diff
@@ -559,14 +570,33 @@ function ahuacatlModifySuite () {
       assertEqual({ a: 1, b: 2 }, actual.json[0].c);
     },
 
+////////////////////////////////////////////////////////////////////////////////
+/// @brief test upsert with no search document
```
Comment (jsteemann) [inline tests/js/server/aql/aql-modify-noncluster.js]: The test actually does something different than the comment says.

File: tests/js/server/aql/aql-modify-noncluster.js [hunk 6]
```diff
@@ -2187,10 +2217,12 @@ function ahuacatlInsertSuite () {
 
 function ahuacatlUpdateSuite () {
   var errors = internal.errors;
-  var cn1 = "UnitTestsAhuacatlUpdate1";
-  var cn2 = "UnitTestsAhuacatlUpdate2";
-  var c1;
-  var c2;
+  const cn1 = "UnitTestsAhuacatlUpdate1";
+  const cn2 = "UnitTestsAhuacatlUpdate2";
+  const cn3 = "UnitTestsAhuacatlUpdate3";
```
Comment (jsteemann) [inline tests/js/server/aql/aql-modify-noncluster.js]: Is it possible to go with one of the already existing collections here, to save test runtime?

File: tests/js/server/aql/aql-modify-noncluster.js [hunk 7]
```diff
@@ -2386,6 +2428,20 @@ function ahuacatlUpdateSuite () {
...
     },
 
+///////////////////////////////////////////////////////////////////////////////
+/// @brief test update with empty object
+////////////////////////////////////////////////////////////////////////////////
+
+    testUpdateEmpty3: function () {
+        const actual = AQL_EXECUTE(`FOR doc IN ${cn3} UPDATE doc WITH {} IN ${cn3} RETURN {old: OLD, new: NEW}`);
+        const res = actual.json;
+        for (let i = 0; i < res.length; ++i) {
```
Comment (jsteemann) [inline tests/js/server/aql/aql-modify-noncluster.js]: Please add an assertion for `res.length` here, so we know that the test actually works on some documents.

**Review by jsteemann — CHANGES_REQUESTED** [review]
See comments

File: arangod/Transaction/Methods.cpp [hunk 8]
```diff
@@ -1510,28 +1514,39 @@ Future<OperationResult> transaction::Methods::modifyLocal(
                             options.returnNew ? &result : nullptr);
     }
 
+    if (result.revisionId() == previous.revisionId()) {
```
[the remark this answers is not in the document]
Comment (cpjulia) [reply arangod/Transaction/Methods.cpp:1517]: Addressed

File: tests/js/server/aql/aql-modify-noncluster.js [hunk 9]
```diff
@@ -141,11 +143,18 @@ function ahuacatlModifySuite () {
     setUp : function () {
       db._drop(cn1);
       db._drop(cn2);
+      db._drop(cn3);
       c1 = db._create(cn1);
       c2 = db._create(cn2);
+      c3 = db._create(cn3);
```
[the remark this answers is not in the document]
Comment (cpjulia) [reply tests/js/server/aql/aql-modify-noncluster.js:149]: Some other test that already existed there is using c2, I could insert more documents into it for my tests, but then would have to restrict the queries in my tests only to the documents I inserted, not the other ones that were being used in the tests that already existed there

File: tests/js/server/aql/aql-modify-noncluster.js (hunk 6 above)
Comment (cpjulia) [reply tests/js/server/aql/aql-modify-noncluster.js:2222]: c1 and c2 were being used by other tests, so I added for separating the documents from my tests

File: tests/js/server/aql/aql-modify-noncluster.js (hunk 7 above)
Comment (cpjulia) [reply tests/js/server/aql/aql-modify-noncluster.js:2438]: Addressed

File: tests/js/server/aql/aql-modify-cluster.js (hunk 3 above)
Comment (cpjulia) [reply tests/js/server/aql/aql-modify-cluster.js:77]: Removed c3

File: arangod/Transaction/Methods.cpp [hunk 10]
```diff
@@ -1077,6 +1074,8 @@ Future<OperationResult> transaction::Methods::insertLocal(
   ManagedDocumentResult docResult;
   ManagedDocumentResult prevDocResult;  // return OLD (with override option)
 
+  size_t numExclusions = 0;
```
Comment (jsteemann) [inline arangod/Transaction/Methods.cpp:1077]:
```suggestion
  [[maybe_unused]] size_t numExclusions = 0;
```

**Review by jsteemann — APPROVED** [review]
LGTM!

**Pull request merged**